### PR TITLE
CloudMonitoring: Allow to set a custom value or disable graph_period

### LINKED ISSUE #48478: Stackdriver: MQL group_by [interval] not supported
**What happened**:

The Stackdriver plugin is appending the `graph_period` to user submitted queries to control the interval of values displayed (value is controlled via the min interval UI or source). This works as long as you aren't setting your own interval in the query via other supported MQL functions. In my example, adding `group_by 1h` to the query returns the 1 hour aggregated value at the interval set by the `graph_period` function, in this case, 20 seconds. 

Example of logging metric with `group_by 1h`
image

**What you expected to happen**:

The time-series interval to be controlled by the `group_by 1h` in the query.

**How to reproduce it (as minimally and precisely as possible)**:
 
- Using the stackdriver datasource
- Switch chart to Edit MQL
- Use any stackdriver metric and add, `| group_by 1h`
- Observe 1 hour aggregated values displayed at interval set by `graph_period`

**Anything else we need to know?**:

I've found that adding an additional `graph_period` to the user submitted query that matches the `group_by` interval correctly displays intended interval.

Example of logging metric with `group_by 1h | graph_period 1h`
image

Another concern I have with the use of `graph_period`, per documentation `graph_period` returns a min, mean, and max value when given a numeric type value, for example, so my question is - which value is being returned? 

**Environment**:
- Grafana version: v8.3.3
- Data source type & version: Stackdriver v1.0

### PR DESCRIPTION
**Pull request by andresmgot**
**What this PR does / why we need it**:

Allow to set a custom value or disable the usage of `graph_period` in Google Cloud Monitoring.

recording

**Which issue(s) this PR fixes**:

Fixes #48478

**Special notes for your reviewer**:

Note that by default it should be enabled (to avoid breaking changes), so that's why people need to specifically disable it.

### CODE REVIEW

File: public/app/plugins/datasource/cloud-monitoring/components/GraphPeriod.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+          </>
+        }
+      >
+        <Switch
+          data-testid={`${refId}-switch-graph-period`}
+          value={graphPeriod !== 'disabled'}
+          onChange={(e) => onChange(e.currentTarget.checked ? '' : 'disabled')}
+        />
+        <Input
+          label="value"
```
Comment (sunker) [inline public/app/plugins/datasource/cloud-monitoring/components/GraphPeriod.tsx:34]: How are they supposed to use this field? Milliseconds? Durations? A tooltip would be helpful.

**Review by sunker — COMMENTED** [review]
Looks good. My only concern is that it's a bit unintuitive that the alignment period field is a dropdown with predefined values whereas this field is an input field. Can we use the same dropdown for graph period? In that case we might want to allow custom values. 

We also need to update the docs. 

But before applying feedback, do we want to run this by our contact persons at google first? I don't have enough insights in MQL to say that https://github.com/grafana/grafana/issues/48478 is a valid ask.

**Comment by andresmgot** [comment]
> But before applying feedback, do we want to run this by our contact persons at google first? I don't have enough insights in MQL to say that #48478 is a valid ask.

As far as I can tell, it does make sense. Setting `graph_period` forces a specific granularity in the data returned. If the data only contains values every minute, we are oversampling with no way of disabling it. We can check anyway if we are right.

File: public/app/plugins/datasource/cloud-monitoring/components/GraphPeriod.test.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  variableOptionGroup: { options: [] },
+};
+
+describe('Graph Period', () => {
+  it('should enable graph_period by default', () => {
+    render(<GraphPeriod {...props} />);
+    expect(screen.getByLabelText('Graph period')).not.toBeDisabled();
+  });
+
+  it('should disable graph_period', async () => {
```
Comment (sarahzinger) [inline public/app/plugins/datasource/cloud-monitoring/components/GraphPeriod.test.tsx:20]:
```suggestion
  it('should disable graph_period when toggled', async () => {
```

File: public/app/plugins/datasource/cloud-monitoring/components/GraphPeriod.test.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    const onChange = jest.fn();
+    render(<GraphPeriod {...props} onChange={onChange} />);
+    const s = screen.getByTestId('A-switch-graph-period');
+    await userEvent.click(s);
+    expect(onChange).toHaveBeenCalledWith('disabled');
+  });
+
+  it('should set a different value', async () => {
```
Comment (sarahzinger) [inline public/app/plugins/datasource/cloud-monitoring/components/GraphPeriod.test.tsx:28]:
```suggestion
  it('should set a different value when selected', async () => {
```

File: public/app/plugins/datasource/cloud-monitoring/types.ts [hunk 4]
```diff
@@ -126,6 +126,7 @@ export interface MetricQuery extends BaseQuery {
   view?: string;
   query: string;
   preprocessor?: PreprocessorType;
+  graphPeriod?: string;
```
Comment (sarahzinger) [inline public/app/plugins/datasource/cloud-monitoring/types.ts:129]: what do you think about graphPeriod: 'disabled' || string or something like that? I could see someone editing this in the future and sending along an empty string thinking they've disabled it when they haven't.

File: pkg/tsdb/cloudmonitoring/time_series_query.go [hunk 5]
```diff
@@ -20,6 +20,19 @@ import (
 	"github.com/grafana/grafana/pkg/tsdb/intervalv2"
 )
 
+func (timeSeriesQuery cloudMonitoringTimeSeriesQuery) appendGraphPeriod(req *backend.QueryDataRequest) string {
+	if timeSeriesQuery.GraphPeriod != "disabled" {
```
Comment (sarahzinger) [inline pkg/tsdb/cloudmonitoring/time_series_query.go:24]: i think a note here explaining that graphPeriod: disabled and graphPeriod: "" are intentionally different could be useful for future versions of ourselves reading this code :D

**Review by sarahzinger — APPROVED** [review]
Looks good I have some minor non-blocking suggestions for you to do with what you will!

**Pull request merged**